tests/unit/s2n_renegotiate_io_test.c

```diff
@@ -0,0 +1,483 @@
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        uint8_t recv_buffer[1] = { 0 };
+
+        /* First handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_SUCCESS(s2n_renegotiate_wipe(client_conn));
+        EXPECT_SUCCESS(s2n_renegotiate_wipe(server_conn));
+
+        ssize_t app_data_read = 0;
+        while (s2n_renegotiate(client_conn, recv_buffer, sizeof(recv_buffer), &app_data_read, &blocked) != S2N_SUCCESS) {
+            EXPECT_EQUAL(s2n_errno, S2N_ERR_IO_BLOCKED);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+            EXPECT_EQUAL(app_data_read, 0);
+            s2n_negotiate(server_conn, &blocked);
+        }
+    }
+
+    /* Test that s2n_renegotiate can handle ApplicationData */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        ssize_t app_data_read = 0;
+        uint8_t recv_buffer[sizeof(app_data)] = { 0 };
+
+        /* First handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_SUCCESS(s2n_renegotiate_wipe(client_conn));
+        EXPECT_SUCCESS(s2n_renegotiate_wipe(server_conn));
+
+        /* Server sends ApplicationData */
+        EXPECT_EQUAL(s2n_send(server_conn, app_data, sizeof(app_data), &blocked), sizeof(app_data));
+
+        /* Client receives ApplicationData */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_renegotiate(client_conn, recv_buffer, sizeof(recv_buffer), &app_data_read, &blocked),
+                S2N_ERR_APP_DATA_BLOCKED);
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_DATA);
+        EXPECT_EQUAL(app_data_read, sizeof(app_data));
+        EXPECT_BYTEARRAY_EQUAL(recv_buffer, app_data, sizeof(app_data));
+
+        /* Client also made progress on the handshake */
+        EXPECT_EQUAL(s2n_conn_get_current_message_type(client_conn), SERVER_HELLO);
```

Comment:
Can you finish the handshake in this test? Stopping at the server hello makes me nervous.